File: src/bus_rl/parity/snapshot.py
"""Deterministic, backend-agnostic snapshots of simulator state.

Snapshots are pure data (NumPy arrays / JSON-compatible scalars) so they can be
written by the Python oracle and, later, produced by the Rust kernel and
compared field by field. Nothing in this module mutates the world.
"""

from __future__ import annotations

import numpy as np

# Fixed layouts. Order is part of the contract; never reorder without bumping
# the fixture schema version.
VEHICLE_FIELDS = (
    "vehicle_id",
    "capacity",
    "route_id",  # -1 == None
    "node",
    "direction",
    "phase",
    "remaining_s",
    "next_node",  # -1 == None
    "visit_id",
    "cooldown_until_s",
    "pattern",
    "turn_stop",  # -1 == None
    "pending_extra",
    "load",
)
COHORT_FIELDS = (
    "cohort_id",
    "lineage_id",
    "route_id",
    "direction",
    "origin_index",
    "destination_index",
    "arrival_tick",
    "count",
    "status",
    "first_denied",
    "boarding_tick",  # -1 == None
    "completion_tick",  # -1 == None
    "abandonment_tick",  # -1 == None
)
COUNTER_FIELDS = (
    "generated",
    "waiting",
    "onboard",
    "completed",
    "abandoned",
    "depot",
)
COST_FIELDS = (
    "waiting_pm",
    "onboard_pm",
    "crowding_pm",
    "active_bus_min",
    "deadhead_bus_min",
    "excessive_wait_pm",
    "first_denied_count",
    "abandoned_count",
    "mission_changes",
    "terminal_unfinished_count",
)
COHORT_KINDS = {"waiting": 0, "onboard": 1, "finished": 2}


def _optional(value: int | None) -> int:
    return -1 if value is None else int(value)


def vehicle_row(vehicle) -> list[int]:
    return [
        vehicle.vehicle_id,
        vehicle.capacity,
        _optional(vehicle.route_id),
        vehicle.node,
        vehicle.direction,
        int(vehicle.phase),
        vehicle.remaining_s,
        _optional(vehicle.next_node),
        vehicle.visit_id,
        vehicle.cooldown_until_s,
        int(vehicle.pattern),
        _optional(vehicle.turn_stop),
        int(vehicle.pending_extra),
        vehicle.load,
    ]


def cohort_row(cohort) -> list[int]:
    return [
        cohort.cohort_id,
        cohort.lineage_id,
        cohort.route_id,
        cohort.direction,
        cohort.origin_index,
        cohort.destination_index,
        cohort.arrival_tick,
        cohort.count,
        int(cohort.status),
        int(cohort.first_denied),
        _optional(cohort.boarding_tick),
        _optional(cohort.completion_tick),
        _optional(cohort.abandonment_tick),
    ]


def counter_array(state) -> np.ndarray:
    return np.array(
        [
            state.generated_count,
            state.waiting_count,
            state.onboard_count,
            state.completed_count,
            state.abandoned_count,
            state.depot_count,
        ],
        dtype=np.int64,
    )


def vehicle_array(state) -> np.ndarray:
    ids = sorted(state.vehicles)
    return np.array([vehicle_row(state.vehicles[i]) for i in ids], dtype=np.int64)


def _cohort_arrays(state) -> tuple[np.ndarray, np.ndarray]:
    rows: list[list[int]] = []
    kinds: list[int] = []
    for cohort in state.cohorts:
        rows.append([-1] + cohort_row(cohort))
        kinds.append(COHORT_KINDS["waiting"])
    for vehicle_id in sorted(state.vehicles):
        for cohort in state.vehicles[vehicle_id].passengers:
            rows.append([vehicle_id] + cohort_row(cohort))
            kinds.append(COHORT_KINDS["onboard"])
    for cohort in state.finished:
        rows.append([-1] + cohort_row(cohort))
        kinds.append(COHORT_KINDS["finished"])
    if not rows:
        return np.zeros((0, len(COHORT_FIELDS) + 1), dtype=np.int64), np.zeros(0, dtype=np.int64)
    return np.array(rows, dtype=np.int64), np.array(kinds, dtype=np.int64)


def numpy_state_snapshot(state) -> dict[str, np.ndarray]:
    """Full world snapshot. Onboard rows are prefixed with the carrying bus id."""
    cohorts, kinds = _cohort_arrays(state)
    routes = sorted(state.headway_targets_s)
    departure_keys = _departure_keys(state)
    return {
        "time_s": np.array(state.current_time_s, dtype=np.int64),
        "counters": counter_array(state),
        "vehicles": vehicle_array(state),
        "cohorts": cohorts,
        "cohort_kind": kinds,
        "headway_targets": np.array([state.headway_targets_s[r] for r in routes], dtype=np.int64),
        "headway_changed_at": np.array(
            [state.headway_changed_at_s[r] for r in routes], dtype=np.int64
        ),
        "terminal_settled": np.array(int(state.terminal_settled), dtype=np.int64),
        "departure_keys": np.array(departure_keys, dtype=np.int64).reshape(-1, 3),
        "last_departure": np.array(
            [state.last_departure_s.get(key, -(10**9)) for key in departure_keys],
            dtype=np.int64,
        ),
        "last_full_departure": np.array(
            [state.last_full_departure_s.get(key, -(10**9)) for key in departure_keys],
            dtype=np.int64,
        ),
    }


def _departure_keys(state) -> list[tuple[int, int, int]]:
    keys = set(state.last_departure_s) | set(state.last_full_departure_s)
    return sorted(keys)


def mask_snapshot(mask) -> np.ndarray:
    return np.asarray(mask, dtype=bool)


def _array_divergence(path: str, expected, actual, rtol: float, atol: float) -> dict | None:
    expected = np.asarray(expected)
    actual = np.asarray(actual)
    if expected.shape != actual.shape:
        return {
            "path": path,
            "reason": "shape",
            "expected": list(expected.shape),
            "actual": list(actual.shape),
        }
    if expected.dtype.kind in "biu" or actual.dtype.kind in "biu":
        equal = np.array_equal(expected, actual)
        if equal:
            return None
        flat = int(np.argmax(np.asarray(expected != actual).reshape(-1)))
    else:
        close = np.isclose(
            expected.astype(np.float64), actual.astype(np.float64), rtol=rtol, atol=atol
        )
        close = close | (np.isnan(expected) & np.isnan(actual))
        if bool(np.all(close)):
            return None
        diff = np.abs(expected.astype(np.float64) - actual.astype(np.float64))
        flat = int(np.argmax(diff.reshape(-1)))
    index = np.unravel_index(flat, expected.shape)
    return {
        "path": f"{path}{[int(i) for i in index]}",
        "expected": expected[index].item() if expected.ndim else expected.item(),
        "actual": actual[index].item() if actual.ndim else actual.item(),
    }


def compare_records(
    expected: dict,
    actual: dict,
    *,
    obs_rtol: float = 1e-6,
    obs_atol: float = 1e-6,
    float_rtol: float = 1e-9,
    float_atol: float = 1e-9,
    limit: int = 20,
) -> list[dict]:
    """Return up to ``limit`` divergences, first one first.

    Exact categories (masks, counters, IDs, status, event lists) use equality.
    Observation tensors use ``obs_*`` tolerance; rewards/costs use the tighter
    cost tolerance from the migration spec.
    """
    divergences: list[dict] = []
    for key in sorted(set(expected) | set(actual)):
        if key not in expected or key not in actual:
            divergences.append(
                {
                    "path": key,
                    "reason": "missing" if key not in actual else "unexpected",
                    "expected": "present" if key in expected else "absent",
                    "actual": "present" if key in actual else "absent",
                }
            )
            if len(divergences) >= limit:
                return divergences
            continue
        left, right = expected[key], actual[key]
        if isinstance(left, dict) or isinstance(right, dict):
            if not isinstance(left, dict) or not isinstance(right, dict):
                divergences.append({"path": key, "reason": "type"})
                continue
            divergences.extend(
                compare_records(
                    left,
                    right,
                    obs_rtol=obs_rtol,
                    obs_atol=obs_atol,
                    float_rtol=float_rtol,
                    float_atol=float_atol,
                    limit=limit - len(divergences),
                )
            )
        elif isinstance(left, list) or isinstance(right, list):
            if left != right:
                divergences.append({"path": key, "expected": left, "actual": right})
        elif "obs" in key:
            obs_key = key.split("__", 1)[1] if "__" in key else key
            found = _array_divergence(f"obs[{obs_key}]", left, right, obs_rtol, obs_atol)
            if found:
                divergences.append(found)
        elif key == "reward" or "cost" in key:
            found = _array_divergence(key, left, right, float_rtol, float_atol)
            if found:
                divergences.append(found)
        else:
            found = _array_divergence(key, left, right, 0.0, 0.0)
            if found:
                divergences.append(found)
        if len(divergences) >= limit:
            break
    return divergences[:limit]


def first_divergence(expected: dict, actual: dict, **kwargs) -> dict | None:
    found = compare_records(expected, actual, limit=1, **kwargs)
    return found[0] if found else None

File: crates/bus-sim-core/src/snapshot.rs
//! Backend-agnostic deterministic snapshots mirroring
//! `bus_sim.parity.snapshot.numpy_state_snapshot`.

use crate::domain::{PassengerCohort, PassengerStatus, Pattern, Phase, Vehicle, WorldState};

pub const COHORT_KIND_WAITING: i64 = 0;
pub const COHORT_KIND_ONBOARD: i64 = 1;
pub const COHORT_KIND_FINISHED: i64 = 2;

pub struct StateSnapshot {
    pub time_s: i64,
    pub counters: [i64; 6],
    pub vehicles: Vec<[i64; 14]>,
    pub cohorts: Vec<[i64; 14]>,
    pub cohort_kind: Vec<i64>,
    pub headway_targets: Vec<i64>,
    pub headway_changed_at: Vec<i64>,
    pub departure_keys: Vec<[i64; 3]>,
    pub last_departure: Vec<i64>,
    pub last_full_departure: Vec<i64>,
    pub terminal_settled: i64,
}

fn optional(value: Option<i64>) -> i64 {
    value.unwrap_or(-1)
}

pub fn vehicle_row(vehicle: &Vehicle) -> [i64; 14] {
    [
        vehicle.vehicle_id as i64,
        vehicle.capacity,
        vehicle.route_id.map(|r| r as i64).unwrap_or(-1),
        vehicle.node as i64,
        vehicle.direction as i64,
        vehicle.phase.as_i64(),
        vehicle.remaining_s,
        vehicle.next_node.map(|n| n as i64).unwrap_or(-1),
        vehicle.visit_id,
        vehicle.cooldown_until_s,
        vehicle.pattern.as_i64(),
        vehicle.turn_stop.map(|t| t as i64).unwrap_or(-1),
        vehicle.pending_extra as i64,
        vehicle.load,
    ]
}

pub fn cohort_row(cohort: &PassengerCohort) -> [i64; 13] {
    [
        cohort.cohort_id,
        cohort.lineage_id,
        cohort.route_id as i64,
        cohort.direction as i64,
        cohort.origin_index as i64,
        cohort.destination_index as i64,
        cohort.arrival_tick,
        cohort.count,
        cohort.status.as_i64(),
        cohort.first_denied as i64,
        optional(cohort.boarding_tick),
        optional(cohort.completion_tick),
        optional(cohort.abandonment_tick),
    ]
}

fn prefixed(bus_id: i64, row: [i64; 13]) -> [i64; 14] {
    [
        bus_id, row[0], row[1], row[2], row[3], row[4], row[5], row[6], row[7], row[8], row[9],
        row[10], row[11], row[12],
    ]
}

pub fn full_snapshot(state: &WorldState) -> StateSnapshot {
    let mut vehicles: Vec<[i64; 14]> = state.vehicles.iter().map(vehicle_row).collect();
    vehicles.sort_by_key(|row| row[0]);

    let mut cohorts: Vec<[i64; 14]> = Vec::new();
    let mut cohort_kind: Vec<i64> = Vec::new();
    for cohort in &state.cohorts {
        cohorts.push(prefixed(-1, cohort_row(cohort)));
        cohort_kind.push(COHORT_KIND_WAITING);
    }
    for vehicle in &state.vehicles {
        for cohort in &vehicle.passengers {
            cohorts.push(prefixed(vehicle.vehicle_id as i64, cohort_row(cohort)));
            cohort_kind.push(COHORT_KIND_ONBOARD);
        }
    }
    for cohort in &state.finished {
        cohorts.push(prefixed(-1, cohort_row(cohort)));
        cohort_kind.push(COHORT_KIND_FINISHED);
    }

    let mut keys: Vec<(u32, i32, u32)> = state.last_departure_s.keys().copied().collect();
    for key in state.last_full_departure_s.keys() {
        if !keys.contains(key) {
            keys.push(*key);
        }
    }
    keys.sort_unstable();

    let departure_keys: Vec<[i64; 3]> = keys
        .iter()
        .map(|(r, d, n)| [*r as i64, *d as i64, *n as i64])
        .collect();
    let last_departure = keys
        .iter()
        .map(|key| *state.last_departure_s.get(key).unwrap_or(&-1_000_000_000))
        .collect();
    let last_full_departure = keys
        .iter()
        .map(|key| {
            *state
                .last_full_departure_s
                .get(key)
                .unwrap_or(&-1_000_000_000)
        })
        .collect();

    StateSnapshot {
        time_s: state.current_time_s,
        counters: [
            state.generated_count(),
            state.waiting_count(),
            state.onboard_count(),
            state.completed_count(),
            state.abandoned_count(),
            state.depot_count(),
        ],
        vehicles,
        cohorts,
        cohort_kind,
        headway_targets: state.headway_targets_s.clone(),
        headway_changed_at: state.headway_changed_at_s.clone(),
        departure_keys,
        last_departure,
        last_full_departure,
        terminal_settled: state.terminal_settled as i64,
    }
}

/// Lightweight per-tick snapshot (vehicles, counters, clock, load) without the
/// full cohort archive, used by the tick-level parity records.
pub struct TickState {
    pub time_s: i64,
    pub counters: [i64; 6],
    pub vehicles: Vec<[i64; 14]>,
}

pub fn tick_snapshot(state: &WorldState) -> TickState {
    TickState {
        time_s: state.current_time_s,
        counters: [
            state.generated_count(),
            state.waiting_count(),
            state.onboard_count(),
            state.completed_count(),
            state.abandoned_count(),
            state.depot_count(),
        ],
        vehicles: state.vehicles.iter().map(vehicle_row).collect(),
    }
}

#[allow(dead_code)]
pub fn pattern_of(value: i64) -> Option<Pattern> {
    match value {
        0 => Some(Pattern::Full),
        1 => Some(Pattern::Short),
        _ => None,
    }
}

#[allow(dead_code)]
pub fn phase_of(value: i64) -> Option<Phase> {
    match value {
        0 => Some(Phase::DepotIdle),
        1 => Some(Phase::Deadhead),
        2 => Some(Phase::TerminalIdle),
        3 => Some(Phase::ServiceMoving),
        4 => Some(Phase::ServiceDwell),
        5 => Some(Phase::Layover),
        _ => None,
    }
}

#[allow(dead_code)]
pub fn status_of(value: i64) -> Option<PassengerStatus> {
    match value {
        0 => Some(PassengerStatus::Waiting),
        1 => Some(PassengerStatus::Onboard),
        2 => Some(PassengerStatus::Completed),
        3 => Some(PassengerStatus::Abandoned),
        _ => None,
    }
}
